src/ElmVersion.hs

```diff
@@ -58,3 +61,22 @@ style_0_19_cannotExposeOpenListing elmVersion =
         Elm_0_18 -> False
         Elm_0_18_Upgrade -> False
         _ -> True
+
+fromFile :: FileStore f => FilePath -> Free f ElmVersion
+fromFile path =
+    do
+        let dir = takeDirectory (path)
+        elmPackageJson <- stat (path </> "elm-package.json")
+        case elmPackageJson of
```

Comment:
Since we are now recursing upward (and it looks like we normalize the path, so we'll recurse up all the way to the root of the filesystem?) so that means we might go up to a folder that contains an `elm.json` or `elm-package.json` that is owned by a different user and maybe not readable by the current user.  Will `stat` still succeed in that case?  Are there any possible permissions that might cause an error? (Maybe we reach a folder that we can't list? though I'm not sure that would be possible given we are in a folder that's inside it.)  If there are possible IO errors from the `stat` due to files that would be outside of the current user's control, I think we ought to make sure we ignore them rather than crashing.

Reply:
Ok. This first naive implementation has at least the merits of raising a few potential design issues, so let's slow down and go back to agreeing on what we want.

First, there does not seem to be an universal agreement of what `normalize` means for paths, but [`System.FilePath.normalise`](https://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:normalise)  used in the PR does not make a path an absolute one. This can be noticed by the fact that it does not use some IO ([`Sytem.Path.toAbsoluteFilePath`](http://hackage.haskell.org/package/paths-0.2.0.0/docs/System-Path.html#v:toAbsoluteFilePath) does). [`System.FilePath.takeDirectory`](http://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:takeDirectory) does not make the path absolute either.

> and it looks like we normalize the path, so we'll recurse up all the way to the root of the filesystem

We don't make the path an absolute path (the only use of `normalise` is in `TestWorld.hs` and this does not make a path absolute as explained previously), so we won't recurse all the way except if the path is already an absolute path. 
`takeDirectory` is only a string manipulation function, it has not real knowledge of the filesystem.
Consequently, for example:
* `./src/file.elm` will stop at `.`
* `/tmp/file/elm` will stop at `/`
* `../test.elm` will stop at `..`
* `..` will actually only test `.` because surprisingly `takeDirectory ".." == "."` :thinking: 

Consequently, if `elm-format `is run from a directory below the `elm.json` or `elm-package.json` file, the version may not be correctly detected. 

For example if in a 0.18 package, `elm-format .` is run from the `src` sub-directory (instead of the package root directory with the `elm-package.json` inside), this will return the default value 0.19.

After your review, it seems to me that preferably, the design should be instead:
* For each file argument:
  - make it absolute (which is actually also an operation that recursively walks upward up to `/`)
  - search for `elm-package.json` or `elm.json` upward up to `/`, trying to detect an upward version
  - then if it is a directory, continue recursively downward to search `elm` files, detecting `elm-package.json` and `elm.json` on the way, and therefore downward versions (in an optimized way), once per directory.

From a performance point of view, this would optimize the `findAllElmFiles` case (a directory argument), but would not optimize the detection of several files arguments, but I think the main use cases are a single file argument or a directory, so this could be acceptable?

A significant downside of this approach is that this would require to really simulate directories in `TestWorld.hs`, which seems quite daunting (also what is the directory above `.` in `TestWorld`?), or to hack a very small working use case for what already exists and do real tests instead for more complicated cases. None seems completely compelling but a compromise would have to be found. Current implementation is simple in `TestWorld` because we don't use absolute paths and only use string operations on paths, which did not require to add filesystem operations in `FileStore.hs`. Maybe I am just lazy :sweat_smile:.

@avh4, what do you think?